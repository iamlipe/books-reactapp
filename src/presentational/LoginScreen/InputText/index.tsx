import React, {memo, forwardRef} from 'react';
import styled, {useTheme} from 'styled-components/native';
import {ReturnKeyTypeOptions, TextInputProps} from 'react-native';
import {useController} from 'react-hook-form';

interface Props extends TextInputProps {
  name: string;
  control: any;
  label: string;
  placeholder: string;
  secureTextEntry?: boolean;
  returnKeyType?: ReturnKeyTypeOptions | undefined;
  onSubmitEditing?: () => void;
  submitButton?: () => JSX.Element;
}

const InputText = forwardRef(
  (
    {
      name,
      control,
      label,
      placeholder,
      secureTextEntry = false,
      onSubmitEditing = () => null,
      returnKeyType = undefined,
      submitButton: SubmitButton,
    }: Props,
    ref,
  ) => {
    const theme = useTheme();

    const {
      field: {onChange, value},
    } = useController({name, control});

    return (
      <StyledContainer>
        <StyledBackground />
        <StyledContainerInput>
          <StyledRow>
            <StyledColumn>
              <StyledLabel>{label}</StyledLabel>
              <StyledInput
                ref={ref}
                placeholder={placeholder}
                placeholderTextColor={theme.colors.TEXT_LIGHT_TRANSPARENT}
                onChangeText={onChange}
                onSubmitEditing={onSubmitEditing}
                secureTextEntry={secureTextEntry}
                returnKeyType={returnKeyType}
                value={value}
              />
            </StyledColumn>
            {SubmitButton ? <SubmitButton /> : null}
          </StyledRow>
        </StyledContainerInput>
      </StyledContainer>
    );
  },
);

const StyledContainer = styled.View`
  width: 100%;
  height: 60px;

  align-self: center;

  margin-bottom: 16px;
`;

const StyledBackground = styled.View`
  position: absolute;
  top: 0;
  left: 0;

  width: 100%;
  height: 100%;

  border-radius: 4px;

  background-color: ${({theme}) => theme.colors.BLACK};
  opacity: 0.32;
`;

const StyledContainerInput = styled.View`
  height: 100%;

  padding: 6px 16px;
`;

const StyledInput = styled.TextInput`
  flex: 1;

  height: 24px;

  color: ${({theme}) => theme.colors.TEXT_LIGHT};

  padding: 0;
`;

const StyledLabel = styled.Text`
  color: ${({theme}) => theme.colors.TEXT_DARK_TRANSPARENT};

  line-height: 16px;
`;

const StyledRow = styled.View`
  height: 100%;

  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const StyledColumn = styled.View`
  min-width: 60%;
`;

export default memo(InputText);
