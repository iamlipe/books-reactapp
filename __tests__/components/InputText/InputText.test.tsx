import React from 'react';
import {Button} from 'react-native';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@config/functions/Testing';

import InputText from '@presentational/LoginScreen/InputText';

const buttonSubmit = () => {
  return <Button testID="button-submit" title="buttonSubmit" />;
};

jest.mock('react-hook-form', () => ({
  useController: () => ({
    field: {
      onChange: jest.fn(),
    },
  }),
}));

describe('InputText', () => {
  const control = jest.fn();
  const name = 'inputName';
  const label = 'inputLabel';
  const placeholder = 'inputPlaceholder';

  test('should render correctly', () => {
    renderWithThemeProvider(
      <InputText
        name={name}
        control={control}
        label={label}
        placeholder={placeholder}
      />,
    );
  });

  test('should match snapshot', () => {
    matchSnapshotWithProvider(
      <InputText
        name={name}
        control={control}
        label={label}
        placeholder={placeholder}
      />,
    );
  });

  test('should render text label', () => {
    const inputText = renderWithThemeProvider(
      <InputText
        name={name}
        control={control}
        label={label}
        placeholder={placeholder}
      />,
    );

    const labelInput = inputText.queryByText(label);

    expect(labelInput).toBeTruthy();
  });

  test('should render placeholder', () => {
    const inputText = renderWithThemeProvider(
      <InputText
        name={name}
        control={control}
        label={label}
        placeholder={placeholder}
      />,
    );

    const input = inputText.getByTestId(`input-text-${name}`);

    expect(input.props.placeholder).toEqual(placeholder);
  });

  test('should render submitButton', () => {
    const inputText = renderWithThemeProvider(
      <InputText
        name={name}
        control={control}
        label={label}
        placeholder={placeholder}
        submitButton={buttonSubmit}
      />,
    );

    const buttonSubmitInput = inputText.queryByTestId('button-submit');

    expect(buttonSubmitInput).toBeTruthy();
  });
});
