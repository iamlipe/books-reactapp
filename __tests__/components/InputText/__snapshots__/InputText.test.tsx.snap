// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "center",
        "height": 60,
        "marginBottom": 16,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#333333",
          "borderRadius": 4,
          "height": "100%",
          "left": 0,
          "opacity": 0.32,
          "position": "absolute",
          "top": 0,
          "width": "100%",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "height": "100%",
          "paddingBottom": 6,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 6,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": "100%",
            "justifyContent": "space-between",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "minWidth": "60%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF52",
                "lineHeight": 16,
              },
            ]
          }
        >
          inputLabel
        </Text>
        <TextInput
          onChangeText={[MockFunction]}
          onSubmitEditing={[Function]}
          placeholder="inputPlaceholder"
          placeholderTextColor="#FFFFFF52"
          secureTextEntry={false}
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "height": 24,
                "paddingBottom": 0,
                "paddingLeft": 0,
                "paddingRight": 0,
                "paddingTop": 0,
              },
            ]
          }
          testID="input-text-inputName"
        />
      </View>
    </View>
  </View>
</View>
`;
